mypy/checker.py

```diff
@@ -1868,27 +1859,53 @@ def visit_call_expr(self, e: CallExpr) -> Type:
-        result = self.accept(e.expr)
+        return_type = self.return_types[-1]
+        subexpr_type = self.accept(e.expr, return_type)
+        if isinstance(subexpr_type, Instance):
+            if subexpr_type.type.fullname() == 'asyncio.futures.Future':
+                # if is a Future, in stmt don't need to do nothing
+                # because the type Future[Some] jus matters to the main loop
+                # that python executes, in statement we shouldn't get the Future,
+                # is just for async purposes.
+                self.function_stack[-1].is_coroutine = True  # Set the function as coroutine
+            elif is_subtype(subexpr_type, self.named_type('typing.Iterable')):
+                # If it's and Iterable-Like, let's check the types.
+                # Maybe just check if have __iter__? (like in analyze_iterable)
+                self.check_iterable_yield_from(e)
+            else:
+                self.msg.yield_from_invalid_operand_type(subexpr_type, e)
+        elif isinstance(subexpr_type, AnyType):
+            self.check_iterable_yield_from(e)
+        else:
+            self.msg.yield_from_invalid_operand_type(subexpr_type, e)
 
-        # If the yield-from isn't typechecked (ie, its Any), don't typecheck
-        if isinstance(result, AnyType):
-            return AnyType()
+        if isinstance(subexpr_type, Instance) and subexpr_type.type.fullname() == 'typing.Generator':
+            return self.get_generator_return_type(subexpr_type)
+        else:
+            # non-Generators don't return anything from "yield from" expressions
+            return Void()
 
-        if not isinstance(result, Instance):
-            self.msg.yield_from_invalid_operand_type(e.expr.accept(self), e)
-        elif result.type.fullname() == "asyncio.futures.Future":
-            self.function_stack[-1].is_coroutine = True  # Set the function as coroutine
-            result = result.args[0]  # Set the return type as the type inside
-        elif is_subtype(result, self.named_type('typing.Iterable')):
-            # TODO
-            # Check return type Iterator[Some]
-            # Maybe set result like in the Future
-            pass
+    def check_iterable_yield_from(self, e: YieldFromExpr) -> Type:
+        """
+            Check that return type is super type of Iterable (Maybe just check if have __iter__?)
+            and compare it with the type of the expression
+        """
+        return_type = self.return_types[-1]
+        expected_item_type = self.get_generator_yield_type(return_type)
+        if e.expr is None:
+            actual_item_type = Void()  # type: Type
         else:
-            # self.msg.yield_from_invalid_operand_type(e.expr, e)
-            self.msg.yield_from_invalid_operand_type(e.expr.accept(self), e)
-        return result
+            actual_item_type = self.accept(e.expr, expected_item_type)
+            if hasattr(actual_item_type, 'args') and cast(Instance, actual_item_type).args:
```

Comment:
It wasn't entirely clear to me what the reasoning behind this is, so I left it alone for now.